lang/LangPrimSource/SC_Ableton_Link.cpp

```diff
@@ -0,0 +1,203 @@
+#ifdef SC_ABLETON_LINK
+
+#include "PyrKernel.h"
+#include "PyrSched.h"
+#include "GC.h"
+#include "PyrPrimitive.h"
+#include "PyrSymbol.h"
+
+#include "SCBase.h"
+#include "SC_Clock.hpp"
+
+#include <ableton/Link.hpp>
+
+static std::chrono::microseconds linkTimeOfInitialization;
+void initLink()
+{
+	linkTimeOfInitialization   = ableton::link::platform::Clock().micros();
+}
+
+inline std::chrono::microseconds hrToLinkTime(double secs){
+	auto time = std::chrono::duration<double>(secs);
+	return std::chrono::duration_cast<std::chrono::microseconds>(time) + linkTimeOfInitialization;
+}
+
+inline double linkToHrTime(std::chrono::microseconds micros){
+	return DurToFloat(micros - linkTimeOfInitialization);
+}
+
+class LinkClock : public TempoClock
+{
+public:
+	LinkClock(VMGlobals *inVMGlobals, PyrObject* inTempoClockObj,
+				double inTempo, double inBaseBeats, double inBaseSeconds);
+
+	~LinkClock() {}
+
+	void SetTempoAtBeat(double inTempo, double inBeats);
+	void SetTempoAtTime(double inTempo, double inSeconds);
+	void SetAll(double inTempo, double inBeats, double inSeconds);
+	double BeatsToSecs(double beats) const
+	{
+		auto timeline = mLink.captureAppTimeline();
+		double secs = linkToHrTime(timeline.timeAtBeat(beats, mQuantum));
+		return secs;
+	}
+	double SecsToBeats(double secs) const
+	{
+		auto timeline = mLink.captureAppTimeline();
+		double beats = timeline.beatAtTime(hrToLinkTime(secs), mQuantum);
+		return beats;
+	}
+
+	void SetQuantum(double quantum);
+	std::size_t NumPeers() const { return mLink.numPeers(); }
+
+private:
+	ableton::Link mLink;
+	double mQuantum;
+};
+
+LinkClock::LinkClock(VMGlobals *inVMGlobals, PyrObject* inTempoClockObj,
+							double inTempo, double inBaseBeats, double inBaseSeconds)
+	: TempoClock(inVMGlobals, inTempoClockObj, inTempo, inBaseBeats, inBaseSeconds),
+	mLink(inTempo * 60.)
+{
+	//quantum = beatsPerBar
+	int err = slotDoubleVal(&inTempoClockObj->slots[2], &mQuantum);
+	if(err) throw err;
+
+	mLink.enable(true);
+	mLink.setTempoCallback([this](double bpm) {
+		double secs = elapsedTime();
+		double tempo = bpm / 60.;
+
+		auto timeline = mLink.captureAppTimeline();
+		double beats = timeline.beatAtTime(hrToLinkTime(secs), mQuantum);
+		mLink.commitAppTimeline(timeline);
```

Comment:
Why `commitAppTimeline` here ? Isn't this method meant to be used to push changes back to Link ? Seems to me at this point no changes need to be committed, the LinkClock is just reading the state from Link.

Reply:
You're right. My proposed fixes remove it: https://github.com/jamshark70/supercollider/commit/f9f38e791881f1568f364b1c9f3ad88922cac4dd

Tested and there is no ill effect.